### DISCUSSION TITLE
How do you implement object properties in native modules?

### DISCUSSION THREAD

**Discussion by lerouxb** [micropython/micropython#13749] [General]
I'm trying to make a native module that has a simple object hierarchy. Imagine the only global it exposes is "Foo"

you'd then use it like:

```python
from my_native_module import Foo

foo = Foo()
foo.bar.some_setter(1)
foo.bar.doSomething()
foo.baz.some_setter(2)
foo.baz.doSomething()
```

so imagine bar and baz in this case are two objects that share a type, each with its own state and some methods.

Looking through the native module documentation and browsing micropython's code I can figure out how to make foo.bar and foo.baz functions or strings or ints, but not other objects. ie. there are lots of examples of achieving foo.bar() and foo.baz(), but what about normal old properties?

Are there any examples somewhere I can copy from? Maybe a project somewhere on the internet that makes more extensive use of native modules?

**Comment by jimmo** [comment]
@lerouxb We don't support implementing properties in native/built-in types.

In general however, we achieve the same result by implementing a custom attribute on the type (via the type's `attr` slot). Look at, for example, `memoryview_attr` in `py/objarray.c` which optionally adds the `itemsize` attribute to `array`.

Similarly in `py/objtype.c` you can see how `__dict__` is implemented.

**Reply by lerouxb** [reply]
Thanks! I suppose in many (most?) cases an alternative pattern would be to create those objects in python land then pass them in to the constructor of the thing that needs them. My example above is perhaps too vague to illustrate it, though. The benefit, I suppose, is that the python code would have a reference to them already without having to look up the attribute.
